### ISSUE TITLE
if x { true } else { false }

### ISSUE THREAD

**Issue by kmcallister** [Manishearth/rust-clippy#12]
Not a Rust-specific problem but I've seen it a lot in code from people who are new to programming in general.

Could generalize to find other forms that are better expressed by logical operators.

**Comment by llogiq** [comment]
I think that one should be simple enough to detect, and it's also easy to create a good warning. I'd catch all instances of if <Expr> { <Lit:bool> } else { <Lit:bool> } and suggest a suitable replacement based on the literals.

I'll take a stab at it when I find some free time.

**Comment by Manishearth** [comment]
Yep, it's pretty simple. Go ahead! :)

**Comment by kmcallister** [comment]
> I'd catch all instances of if { Lit:bool } else { Lit:bool } and suggest a suitable replacement based on the literals.

It'd be great to have a general, declarative way to express rules like that. Maybe something similar to [Coccinelle](http://coccinelle.lip6.fr/).

**Comment by Manishearth** [comment]
Technically it's just a single huge match statement (assuming `syntax::ptr::P` doesn't get in the way) ;)

**Comment by kmcallister** [comment]
It would be cool to do this with quotation:

```rust
fn check_expr(&mut self, cx: &Context, expr: &Expr) {
    match *expr {
        q!(if $e:expr { $a:lit } else { $b:lit }) => ...
```

I think this is actually achievable now, if you replace the whole `match` by an invocation of a procedural macro, similar to html5ever's `match_token!`.

**Comment by llogiq** [comment]
A macro should be able to generate code to deal with `ptr::P`, as irritating as it sometimes is. But until we have this macro, I'm going to write the lint the traditional way.

**Comment by llogiq** [comment]
Done. The warnings could give more specific directions (they are a bit abstract for a newbie), but it works, and it's late now; I guess, I'll revisit them later.
